numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -755,66 +756,193 @@ VOID_getitem(void *input, void *vap)
+                PyErr_SetString(PyExc_TypeError, msg_nostruct);
+                return -1;
+            }
 
-        res = 0;
-        /* get the names from the fields dictionary*/
-        names = descr->names;
-        n = PyTuple_GET_SIZE(names);
-        if (PyTuple_GET_SIZE(op) != n) {
-            PyErr_SetString(PyExc_ValueError,
-                    "size of tuple must match number of fields.");
-            return -1;
+            srcdescr = PyArray_DESCR(oparr);
+            srcdata = PyArray_DATA(oparr);
         }
-        savedflags = PyArray_FLAGS(ap);
-        for (i = 0; i < n; i++) {
-            PyArray_Descr *new;
-            npy_intp offset;
-            key = PyTuple_GET_ITEM(names, i);
-            tup = PyDict_GetItem(descr->fields, key);
-            if (_unpack_field(tup, &new, &offset) < 0) {
-                ((PyArrayObject_fields *)ap)->descr = descr;
+        else if (PyArray_IsScalar(op, Void)) {
+            srcdescr = ((PyVoidScalarObject*)op)->descr;
+            if (!PyDataType_HASFIELDS(srcdescr)) {
+                PyErr_SetString(PyExc_TypeError, msg_nostruct);
                 return -1;
             }
-            /*
-             * TODO: temporarily modifying the array like this
-             *       is bad coding style, should be changed.
-             */
-            ((PyArrayObject_fields *)ap)->descr = new;
-            /* remember to update alignment flags */
-            if ((new->alignment > 1)
-                    && ((((npy_intp)(ip+offset)) % new->alignment) != 0)) {
-                PyArray_CLEARFLAGS(ap, NPY_ARRAY_ALIGNED);
+            srcdata = ((PyVoidScalarObject*)op)->obval;
+        }
+
+        names_size = PyTuple_GET_SIZE(descr->names);
+
+        /* if we found the src dtype above, simply copy and return */
+        if (srcdescr != NULL) {
+            int ret;
+
+            /* Fast path if dtypes are equal */
+            if (PyArray_EquivTypes(srcdescr, descr)) {
+                for (i = 0; i < names_size; i++) {
+                    /* neither line can ever fail, in principle */
+                    if (_setupField(i, descr, dummy, &offset)) {
+                        return -1;
+                    }
+                    PyArray_DESCR(dummy)->f->copyswap(ip + offset,
+                                                 srcdata + offset, 0, dummy);
+                }
+                return 0;
             }
-            else {
-                PyArray_ENABLEFLAGS(ap, NPY_ARRAY_ALIGNED);
+
+            /* Slow path */
+            ret = PyArray_CastRawArrays(1, srcdata, ip, 0, 0,
+                                        srcdescr, descr, 0);
+            if (ret != NPY_SUCCEED) {
+                return -1;
             }
-            res = new->f->setitem(PyTuple_GET_ITEM(op, i), ip+offset, ap);
-            ((PyArrayObject_fields *)ap)->flags = savedflags;
-            if (res < 0) {
-                break;
+            return 0;
+        }
```

Comment:
This `if` statement might read better as a helper function as then it can appear in the `PyArray_Check(op)` and `PyArray_IsScalar`  branches above, and the whole function becomes a single chained if/else